File: NextcloudTalk/Maps/MapViewController.m
//
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "MapViewController.h"

#import <CoreLocation/CoreLocation.h>

#import "GeoLocationRichObject.h"
#import "NCAppBranding.h"

@interface MapViewController () <CLLocationManagerDelegate, MKMapViewDelegate>
{
    CLLocationManager *_locationManager;
    MKPointAnnotation *_annotation;
}

@end

@implementation MapViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    [NCAppBranding styleViewController:self];

    self.navigationItem.title = NSLocalizedString(@"Shared location", nil);

    _locationManager = [[CLLocationManager alloc] init];
    _locationManager.delegate = self;
    [_locationManager requestWhenInUseAuthorization];
    
    _mapView.delegate = self;
    [self centerMapViewInSharedLocation];
    [_mapView addAnnotation:_annotation];
    
    UIBarButtonItem *cancelButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
                                                                                  target:self action:@selector(cancelButtonPressed)];
    self.navigationItem.leftBarButtonItem = cancelButton;

    UIImage *shareImage = [UIImage systemImageNamed:@"square.and.arrow.up"];
    UIBarButtonItem *shareButton = [[UIBarButtonItem alloc] initWithImage:shareImage style:UIBarButtonItemStylePlain
                                                                   target:nil action:nil];
    shareButton.menu = [self shareOptionsMenu];
    self.navigationItem.rightBarButtonItem = shareButton;
}

- (instancetype)initWithGeoLocationRichObject:(GeoLocationRichObject *)geoLocation
{
    self = [super init];
    if (self) {
        _annotation = [[MKPointAnnotation alloc] init];
        _annotation.coordinate = CLLocationCoordinate2DMake([geoLocation.latitude doubleValue], [geoLocation.longitude doubleValue]);
        _annotation.title = geoLocation.name;
    }
    
    return self;
}

#pragma mark - Actions

- (void)cancelButtonPressed
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (UIMenu *)shareOptionsMenu
{
    UIAction *appleMapsAction = [UIAction actionWithTitle:NSLocalizedString(@"Open in Maps", nil)
                                                    image:nil
                                               identifier:nil
                                                  handler:^(UIAction *action) {
        [self openInMaps:self->_annotation.coordinate];
    }];

    UIAction *googleMapsAction = nil;

    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"comgooglemaps://"]]) {
        googleMapsAction = [UIAction actionWithTitle:NSLocalizedString(@"Open in Google Maps", nil)
                                               image:nil
                                          identifier:nil
                                             handler:^(UIAction *action) {
            [self openInGoogleMaps:self->_annotation.coordinate];
        }];
    }

    NSMutableArray *actions = [NSMutableArray arrayWithObject:appleMapsAction];
    if (googleMapsAction) {
        [actions addObject:googleMapsAction];
    }

    return [UIMenu menuWithTitle:@"" children:actions];
}


#pragma mark - Map view

- (void)centerMapViewInSharedLocation
{
    MKCoordinateRegion mapRegion;
    mapRegion.center = _annotation.coordinate;
    mapRegion.span = MKCoordinateSpanMake(0.005, 0.005);
    [self.mapView setRegion:mapRegion animated:YES];
}

#pragma mark - Utils

- (void)openInMaps:(CLLocationCoordinate2D)coordinate {
    MKPlacemark *placemark = [[MKPlacemark alloc] initWithCoordinate:coordinate];
    MKMapItem *mapItem = [[MKMapItem alloc] initWithPlacemark:placemark];
    mapItem.name = NSLocalizedString(@"Shared location", nil);

    NSDictionary *options = @{
        MKLaunchOptionsMapCenterKey: [NSValue valueWithMKCoordinate:coordinate]
    };

    [mapItem openInMapsWithLaunchOptions:options];
}

- (void)openInGoogleMaps:(CLLocationCoordinate2D)coordinate {
    NSString *urlString = [NSString stringWithFormat:
                           @"comgooglemaps://?q=%f,%f&center=%f,%f&zoom=14",
                           coordinate.latitude,
                           coordinate.longitude,
                           coordinate.latitude,
                           coordinate.longitude];

    NSURL *url = [NSURL URLWithString:urlString];

    if ([[UIApplication sharedApplication] canOpenURL:url]) {
        [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
    }
}

#pragma mark - MKMapViewDelegate

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id <MKAnnotation>)annotation
{
    // If the annotation is the user location, just return nil.
    if ([annotation isKindOfClass:[MKUserLocation class]]) {
        return nil;
    }
    
    if (annotation == _annotation) {
        MKPinAnnotationView *pinView = (MKPinAnnotationView*)[mapView dequeueReusableAnnotationViewWithIdentifier:@"SharedLocationAnnotationView"];
        if (!pinView) {
            pinView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"SharedLocationAnnotationView"];
            pinView.pinTintColor = [NCAppBranding elementColor];
            pinView.animatesDrop = YES;
            pinView.canShowCallout = YES;
        } else {
            pinView.annotation = annotation;
        }
        return pinView;
    }
    return nil;
}

@end

File: NextcloudTalk/Maps/MapViewController.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import UIKit
import CoreLocation
import MapKit

class MapViewController: UIViewController,
                         CLLocationManagerDelegate,
                         MKMapViewDelegate {

    private static let annotationReuseIdentifier = "SharedLocationAnnotationView"

    @IBOutlet weak var mapView: MKMapView!

    private var locationManager: CLLocationManager!
    private let sharedLocationAnnotation: MKPointAnnotation

    init(geoLocationRichObject geoLocation: GeoLocationRichObject) {
        self.sharedLocationAnnotation = MKPointAnnotation()
        self.sharedLocationAnnotation.coordinate = CLLocationCoordinate2D(latitude: Double(geoLocation.latitude) ?? 0,
                                                                         longitude: Double(geoLocation.longitude) ?? 0)
        self.sharedLocationAnnotation.title = geoLocation.name

        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        NCAppBranding.styleViewController(self)

        self.navigationItem.title = NSLocalizedString("Shared location", comment: "")

        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.requestWhenInUseAuthorization()

        self.mapView.delegate = self
        self.centerMapViewInSharedLocation()
        self.mapView.addAnnotation(self.sharedLocationAnnotation)

        self.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel,
                                                                target: self,
                                                                action: #selector(cancelButtonPressed))

        let shareButton = UIBarButtonItem(image: UIImage(systemName: "square.and.arrow.up"), style: .plain, target: nil, action: nil)
        shareButton.menu = self.shareOptionsMenu()
        self.navigationItem.rightBarButtonItem = shareButton
    }

    // MARK: - Actions

    @objc func cancelButtonPressed() {
        self.dismiss(animated: true)
    }

    private func shareOptionsMenu() -> UIMenu {
        var actions: [UIAction] = []

        actions.append(UIAction(title: NSLocalizedString("Open in Maps", comment: ""), image: nil, identifier: nil) { [weak self] _ in
            guard let self else { return }

            self.openInMaps(self.sharedLocationAnnotation.coordinate)
        })

        if let googleMapsUrl = URL(string: "comgooglemaps://"), UIApplication.shared.canOpenURL(googleMapsUrl) {
            actions.append(UIAction(title: NSLocalizedString("Open in Google Maps", comment: ""), image: nil, identifier: nil) { [weak self] _ in
                guard let self else { return }

                self.openInGoogleMaps(self.sharedLocationAnnotation.coordinate)
            })
        }

        return UIMenu(title: "", children: actions)
    }

    // MARK: - Map view

    private func centerMapViewInSharedLocation() {
        let mapRegion = MKCoordinateRegion(center: self.sharedLocationAnnotation.coordinate,
                                           span: MKCoordinateSpan(latitudeDelta: 0.005, longitudeDelta: 0.005))

        self.mapView.setRegion(mapRegion, animated: true)
    }

    // MARK: - Utils

    private func openInMaps(_ coordinate: CLLocationCoordinate2D) {
        let mapItem = MKMapItem(placemark: MKPlacemark(coordinate: coordinate))
        mapItem.name = NSLocalizedString("Shared location", comment: "")

        mapItem.openInMaps(launchOptions: [MKLaunchOptionsMapCenterKey: NSValue(mkCoordinate: coordinate)])
    }

    private func openInGoogleMaps(_ coordinate: CLLocationCoordinate2D) {
        let urlString = String(format: "comgooglemaps://?q=%f,%f&center=%f,%f&zoom=14",
                               coordinate.latitude, coordinate.longitude,
                               coordinate.latitude, coordinate.longitude)

        guard let url = URL(string: urlString), UIApplication.shared.canOpenURL(url) else { return }

        UIApplication.shared.open(url)
    }

    // MARK: - MKMapViewDelegate

    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        // If the annotation is the user location, just return nil.
        if annotation is MKUserLocation {
            return nil
        }

        guard annotation === self.sharedLocationAnnotation else { return nil }

        if let pinView = mapView.dequeueReusableAnnotationView(withIdentifier: Self.annotationReuseIdentifier) as? MKPinAnnotationView {
            pinView.annotation = annotation

            return pinView
        }

        let pinView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: Self.annotationReuseIdentifier)
        pinView.pinTintColor = NCAppBranding.elementColor()
        pinView.animatesDrop = true
        pinView.canShowCallout = true

        return pinView
    }
}
